File: server/infra/telemetry/reader.py
"""Prometheus exposition format renderer — read metrics from Redis and format.

Responsibilities:
1. `.` → `_` name conversion
2. Type suffix: counter → `_total`, histogram unit=s → `_seconds`
3. `# HELP` / `# TYPE` from {otlp}:meta
4. Inject `service` label from {otlp}:resource
5. Histogram expansion: `_bucket{le=...}`, `_sum`, `_count`
   - Buckets are converted from per-bucket delta counts to cumulative counts
6. Gauge aggregation: SMEMBERS → HGETALL per worker → SUM, lazy SREM
7. Label value escaping per Prometheus spec
8. Content-Type: text/plain; version=0.0.4; charset=utf-8
"""

from __future__ import annotations

import json
from collections import defaultdict

from redis.asyncio import Redis, RedisCluster

from infra.telemetry.redis_exporter import (
    KEY_COUNTERS,
    KEY_HISTOGRAMS,
    KEY_GAUGE_PIDS,
    KEY_META,
    KEY_RESOURCE,
    _gauge_key,
)


def _escape_label_value(v: str) -> str:
    """Escape label value per Prometheus exposition spec."""
    return v.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")


def _prom_name(otel_name: str) -> str:
    """Convert OTel dot-separated name to Prometheus underscore-separated."""
    return otel_name.replace(".", "_")


def _format_labels(labels: dict[str, str]) -> str:
    """Format label dict to Prometheus label string: {k1="v1",k2="v2"}"""
    if not labels:
        return ""
    parts = ",".join(
        f'{k}="{_escape_label_value(v)}"' for k, v in sorted(labels.items())
    )
    return "{" + parts + "}"


def _format_value(v: float) -> str:
    """Format a numeric value, omitting unnecessary decimals for integers."""
    if v != v:  # NaN check (NaN != NaN)
        return "NaN"
    if v == float("inf"):
        return "+Inf"
    if v == float("-inf"):
        return "-Inf"
    if v == int(v):
        return str(int(v))
    return str(v)


def _parse_field(field: str) -> tuple[str, dict] | None:
    """Parse a Redis field key: 'metric_name|{attrs_json}' → (name, attrs_dict).

    Returns None if the field is malformed.
    """
    try:
        sep = field.index("|")
        name = field[:sep]
        attrs = json.loads(field[sep + 1:])
        return name, attrs
    except (ValueError, json.JSONDecodeError):
        return None


async def render_prometheus_exposition(redis: Redis | RedisCluster) -> str:
    """Read all metrics from Redis and render Prometheus exposition text."""
    lines: list[str] = []

    # ── Load resource info ──
    service_name = await redis.hget(KEY_RESOURCE, "service.name") or ""

    # ── Load metadata ──
    meta_raw: dict = await redis.hgetall(KEY_META) or {}
    meta: dict[str, dict] = {}
    for k, v in meta_raw.items():
        try:
            meta[k] = json.loads(v)
        except (json.JSONDecodeError, TypeError):
            pass

    # ── Counters ──
    counters_raw: dict = await redis.hgetall(KEY_COUNTERS) or {}
    # Group by metric name
    counter_groups: dict[str, list[tuple[dict, float]]] = defaultdict(list)
    for field, val in counters_raw.items():
        parsed = _parse_field(field)
        if not parsed:
            continue
        name, attrs = parsed
        counter_groups[name].append((attrs, float(val)))

    for otel_name in sorted(counter_groups):
        pname = _prom_name(otel_name) + "_total"
        m = meta.get(otel_name, {})
        lines.append(f"# HELP {pname} {m.get('description', '')}")
        lines.append(f"# TYPE {pname} counter")
        for attrs, value in counter_groups[otel_name]:
            labels = _inject_service(attrs, service_name)
            lines.append(f"{pname}{_format_labels(labels)} {_format_value(value)}")
        lines.append("")

    # ── Histograms ──
    hist_raw: dict = await redis.hgetall(KEY_HISTOGRAMS) or {}
    # Parse into structure: {metric_name: {attrs_json: {suffix: value}}}
    hist_data: dict[str, dict[str, dict[str, float]]] = defaultdict(
        lambda: defaultdict(dict)
    )
    for field, val in hist_raw.items():
        # field = "metric_name|{attrs_json}|suffix"
        # Use rsplit to split from the RIGHT — suffix (count/sum/bucket_*) never
        # contains "|", so this is safe even if attrs_json contains pipe chars.
        parts = field.rsplit("|", 1)
        if len(parts) != 2:
            continue
        base_key, suffix = parts
        # base_key = "metric_name|{attrs_json}"
        parsed = _parse_field(base_key)
        if not parsed:
            continue
        name, attrs = parsed
        attrs_json = json.dumps(dict(sorted(attrs.items())), ensure_ascii=False)
        hist_data[name][attrs_json][suffix] = float(val)

    for otel_name in sorted(hist_data):
        m = meta.get(otel_name, {})
        unit = m.get("unit", "")
        pname = _prom_name(otel_name)
        if unit == "s":
            pname += "_seconds"

        lines.append(f"# HELP {pname} {m.get('description', '')}")
        lines.append(f"# TYPE {pname} histogram")

        for attrs_json in sorted(hist_data[otel_name]):
            attrs = json.loads(attrs_json)
            labels = _inject_service(attrs, service_name)
            buckets = hist_data[otel_name][attrs_json]

            count = buckets.get("count", 0)
            total_sum = buckets.get("sum", 0)

            # Collect bucket boundaries and sort numerically
            bucket_bounds: list[tuple[float | str, float]] = []
            for k, v in buckets.items():
                if k.startswith("bucket_"):
                    bound_str = k[7:]  # strip "bucket_"
                    if bound_str == "+Inf":
                        bucket_bounds.append((float("inf"), v))
                    else:
                        bucket_bounds.append((float(bound_str), v))

            bucket_bounds.sort(key=lambda x: x[0])

            # Convert from per-bucket (delta) counts to cumulative counts
            cumulative = 0.0
            for bound, delta_count in bucket_bounds:
                cumulative += delta_count
                le_str = "+Inf" if bound == float("inf") else _format_value(bound)
                blabels = {**labels, "le": le_str}
                lines.append(
                    f"{pname}_bucket{_format_labels(blabels)} {_format_value(cumulative)}"
                )

            lines.append(f"{pname}_sum{_format_labels(labels)} {_format_value(total_sum)}")
            lines.append(f"{pname}_count{_format_labels(labels)} {_format_value(count)}")
        lines.append("")

    # ── Gauges (UpDownCounter — per-worker aggregation) ──
    gauge_pids: set = await redis.smembers(KEY_GAUGE_PIDS) or set()
    # Aggregate across all alive workers
    gauge_agg: dict[str, float] = defaultdict(float)
    dead_pids: list[str] = []

    for pid in gauge_pids:
        gk = _gauge_key(pid)
        fields: dict = await redis.hgetall(gk) or {}
        if not fields:
            # Key expired — worker crashed; lazy cleanup
            dead_pids.append(pid)
            continue
        for field, val in fields.items():
            gauge_agg[field] += float(val)

    # Lazy SREM dead PIDs
    if dead_pids:
        await redis.srem(KEY_GAUGE_PIDS, *dead_pids)

    # Group by metric name
    gauge_groups: dict[str, list[tuple[dict, float]]] = defaultdict(list)
    for field, total in gauge_agg.items():
        parsed = _parse_field(field)
        if not parsed:
            continue
        name, attrs = parsed
        gauge_groups[name].append((attrs, total))

    for otel_name in sorted(gauge_groups):
        pname = _prom_name(otel_name)
        m = meta.get(otel_name, {})
        lines.append(f"# HELP {pname} {m.get('description', '')}")
        lines.append(f"# TYPE {pname} gauge")
        for attrs, value in gauge_groups[otel_name]:
            labels = _inject_service(attrs, service_name)
            lines.append(f"{pname}{_format_labels(labels)} {_format_value(value)}")
        lines.append("")

    return "\n".join(lines)


async def reset_all_metrics(redis: Redis | RedisCluster) -> None:
    """Delete all OTLP metric keys from Redis."""
    keys_to_del = [KEY_COUNTERS, KEY_HISTOGRAMS, KEY_META, KEY_RESOURCE]

    # Collect gauge per-worker keys
    gauge_pids: set = await redis.smembers(KEY_GAUGE_PIDS) or set()
    for pid in gauge_pids:
        keys_to_del.append(_gauge_key(pid))
    keys_to_del.append(KEY_GAUGE_PIDS)

    if keys_to_del:
        await redis.delete(*keys_to_del)


def _inject_service(attrs: dict, service_name: str) -> dict[str, str]:
    """Return a new label dict with service injected as the first key."""
    labels: dict[str, str] = {}
    if service_name:
        labels["service"] = service_name
    labels.update({k: str(v) for k, v in attrs.items()})
    return labels

File: backend/internal/infra/telemetry/reader.go
package telemetry

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strings"

	"github.com/redis/go-redis/v9"
	"github.com/rs/zerolog/log"
)

func escapeLabel(v string) string {
	v = strings.ReplaceAll(v, `\`, `\\`)
	v = strings.ReplaceAll(v, `"`, `\"`)
	v = strings.ReplaceAll(v, "\n", `\n`)
	return v
}

func promName(otelName string) string {
	return strings.ReplaceAll(otelName, ".", "_")
}

func formatLabels(labels map[string]string) string {
	if len(labels) == 0 {
		return ""
	}
	keys := make([]string, 0, len(labels))
	for k := range labels {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, len(keys))
	for i, k := range keys {
		parts[i] = fmt.Sprintf(`%s="%s"`, k, escapeLabel(labels[k]))
	}
	return "{" + strings.Join(parts, ",") + "}"
}

func formatValue(v float64) string {
	if math.IsNaN(v) {
		return "NaN"
	}
	if math.IsInf(v, 1) {
		return "+Inf"
	}
	if math.IsInf(v, -1) {
		return "-Inf"
	}
	if v == float64(int64(v)) {
		return fmt.Sprintf("%d", int64(v))
	}
	return fmt.Sprintf("%v", v)
}

func parseField(field string) (string, map[string]string, bool) {
	idx := strings.Index(field, "|")
	if idx < 0 {
		return "", nil, false
	}
	name := field[:idx]
	var attrs map[string]string
	if err := json.Unmarshal([]byte(field[idx+1:]), &attrs); err != nil {
		return "", nil, false
	}
	return name, attrs, true
}

func injectService(attrs map[string]string, serviceName string) map[string]string {
	labels := make(map[string]string)
	if serviceName != "" {
		labels["service"] = serviceName
	}
	for k, v := range attrs {
		labels[k] = v
	}
	return labels
}

// RenderPrometheusExposition reads all metrics from Redis and renders Prometheus text.
func RenderPrometheusExposition(ctx context.Context, rdb redis.UniversalClient) (string, error) {
	var lines []string

	// Load resource info
	serviceName, err := rdb.HGet(ctx, KeyResource, "service.name").Result()
	if err != nil && err != redis.Nil {
		log.Warn().Err(err).Msg("telemetry: failed to read resource info")
	}

	// Load metadata
	metaRaw, err := rdb.HGetAll(ctx, KeyMeta).Result()
	if err != nil {
		log.Warn().Err(err).Msg("telemetry: failed to read metadata")
	}
	meta := make(map[string]map[string]string)
	for k, v := range metaRaw {
		var m map[string]string
		if json.Unmarshal([]byte(v), &m) == nil {
			meta[k] = m
		}
	}

	// Counters
	countersRaw, err := rdb.HGetAll(ctx, KeyCounters).Result()
	if err != nil {
		log.Warn().Err(err).Msg("telemetry: failed to read counters")
	}
	counterGroups := make(map[string][]struct {
		attrs map[string]string
		value float64
	})
	for field, val := range countersRaw {
		name, attrs, ok := parseField(field)
		if !ok {
			continue
		}
		v := parseFloat(val)
		counterGroups[name] = append(counterGroups[name], struct {
			attrs map[string]string
			value float64
		}{attrs, v})
	}

	sortedCounters := sortedKeys(counterGroups)
	for _, otelName := range sortedCounters {
		pname := promName(otelName) + "_total"
		m := meta[otelName]
		lines = append(lines, fmt.Sprintf("# HELP %s %s", pname, getOr(m, "description", "")))
		lines = append(lines, fmt.Sprintf("# TYPE %s counter", pname))
		for _, entry := range counterGroups[otelName] {
			labels := injectService(entry.attrs, serviceName)
			lines = append(lines, fmt.Sprintf("%s%s %s", pname, formatLabels(labels), formatValue(entry.value)))
		}
		lines = append(lines, "")
	}

	// Histograms
	histRaw, err := rdb.HGetAll(ctx, KeyHistograms).Result()
	if err != nil {
		log.Warn().Err(err).Msg("telemetry: failed to read histograms")
	}
	type histEntry struct {
		suffixes map[string]float64
	}
	histData := make(map[string]map[string]*histEntry) // name -> attrsJSON -> entry

	for field, val := range histRaw {
		parts := splitFromRight(field, "|")
		if len(parts) != 2 {
			continue
		}
		baseKey, suffix := parts[0], parts[1]
		name, attrs, ok := parseField(baseKey)
		if !ok {
			continue
		}
		attrsJSON, _ := json.Marshal(attrs)
		ak := string(attrsJSON)

		if histData[name] == nil {
			histData[name] = make(map[string]*histEntry)
		}
		if histData[name][ak] == nil {
			histData[name][ak] = &histEntry{suffixes: make(map[string]float64)}
		}
		histData[name][ak].suffixes[suffix] = parseFloat(val)
	}

	sortedHists := sortedKeys(histData)
	for _, otelName := range sortedHists {
		m := meta[otelName]
		pname := promName(otelName)
		if getOr(m, "unit", "") == "s" {
			pname += "_seconds"
		}
		lines = append(lines, fmt.Sprintf("# HELP %s %s", pname, getOr(m, "description", "")))
		lines = append(lines, fmt.Sprintf("# TYPE %s histogram", pname))

		attrsKeys := sortedKeys(histData[otelName])
		for _, ak := range attrsKeys {
			var attrs map[string]string
			json.Unmarshal([]byte(ak), &attrs)
			labels := injectService(attrs, serviceName)
			buckets := histData[otelName][ak].suffixes

			count := buckets["count"]
			totalSum := buckets["sum"]

			// Collect and sort bucket boundaries
			type bucketBound struct {
				bound float64
				count float64
			}
			var bounds []bucketBound
			for k, v := range buckets {
				if strings.HasPrefix(k, "bucket_") {
					boundStr := k[7:]
					if boundStr == "+Inf" {
						bounds = append(bounds, bucketBound{math.Inf(1), v})
					} else {
						bounds = append(bounds, bucketBound{parseFloat(boundStr), v})
					}
				}
			}
			sort.Slice(bounds, func(i, j int) bool { return bounds[i].bound < bounds[j].bound })

			// Convert delta to cumulative
			cumulative := 0.0
			for _, b := range bounds {
				cumulative += b.count
				leStr := "+Inf"
				if !math.IsInf(b.bound, 1) {
					leStr = formatValue(b.bound)
				}
				blabels := copyMap(labels)
				blabels["le"] = leStr
				lines = append(lines, fmt.Sprintf("%s_bucket%s %s", pname, formatLabels(blabels), formatValue(cumulative)))
			}
			lines = append(lines, fmt.Sprintf("%s_sum%s %s", pname, formatLabels(labels), formatValue(totalSum)))
			lines = append(lines, fmt.Sprintf("%s_count%s %s", pname, formatLabels(labels), formatValue(count)))
		}
		lines = append(lines, "")
	}

	// Gauges
	gaugePIDs, err := rdb.SMembers(ctx, KeyGaugePIDs).Result()
	if err != nil {
		log.Warn().Err(err).Msg("telemetry: failed to read gauge PIDs")
	}
	gaugeAgg := make(map[string]float64)
	var deadPIDs []string

	for _, pid := range gaugePIDs {
		gk := GaugeKey(pid)
		fields, err := rdb.HGetAll(ctx, gk).Result()
		if err != nil {
			log.Warn().Err(err).Str("pid", pid).Msg("telemetry: failed to read gauge data")
		}
		if len(fields) == 0 {
			deadPIDs = append(deadPIDs, pid)
			continue
		}
		for field, val := range fields {
			gaugeAgg[field] += parseFloat(val)
		}
	}
	if len(deadPIDs) > 0 {
		rdb.SRem(ctx, KeyGaugePIDs, toInterfaces(deadPIDs)...)
	}

	gaugeGroups := make(map[string][]struct {
		attrs map[string]string
		value float64
	})
	for field, total := range gaugeAgg {
		name, attrs, ok := parseField(field)
		if !ok {
			continue
		}
		gaugeGroups[name] = append(gaugeGroups[name], struct {
			attrs map[string]string
			value float64
		}{attrs, total})
	}

	sortedGauges := sortedKeys(gaugeGroups)
	for _, otelName := range sortedGauges {
		pname := promName(otelName)
		m := meta[otelName]
		lines = append(lines, fmt.Sprintf("# HELP %s %s", pname, getOr(m, "description", "")))
		lines = append(lines, fmt.Sprintf("# TYPE %s gauge", pname))
		for _, entry := range gaugeGroups[otelName] {
			labels := injectService(entry.attrs, serviceName)
			lines = append(lines, fmt.Sprintf("%s%s %s", pname, formatLabels(labels), formatValue(entry.value)))
		}
		lines = append(lines, "")
	}

	return strings.Join(lines, "\n"), nil
}

// ResetAllMetrics deletes all OTLP metric keys from Redis.
func ResetAllMetrics(ctx context.Context, rdb redis.UniversalClient) error {
	keys := []string{KeyCounters, KeyHistograms, KeyMeta, KeyResource}
	pids, err := rdb.SMembers(ctx, KeyGaugePIDs).Result()
	if err != nil {
		log.Warn().Err(err).Msg("telemetry: failed to read gauge PIDs for reset")
	}
	for _, pid := range pids {
		keys = append(keys, GaugeKey(pid))
	}
	keys = append(keys, KeyGaugePIDs)
	return rdb.Del(ctx, keys...).Err()
}

// Helper functions
func parseFloat(s string) float64 {
	v := 0.0
	fmt.Sscanf(s, "%f", &v)
	return v
}

func splitFromRight(s, sep string) []string {
	idx := strings.LastIndex(s, sep)
	if idx < 0 {
		return []string{s}
	}
	return []string{s[:idx], s[idx+1:]}
}

func getOr(m map[string]string, key, def string) string {
	if m == nil {
		return def
	}
	if v, ok := m[key]; ok {
		return v
	}
	return def
}

func copyMap(m map[string]string) map[string]string {
	c := make(map[string]string, len(m))
	for k, v := range m {
		c[k] = v
	}
	return c
}

func sortedKeys[V any](m map[string]V) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

func toInterfaces(ss []string) []interface{} {
	out := make([]interface{}, len(ss))
	for i, s := range ss {
		out[i] = s
	}
	return out
}
